File: src/main/java/algo/depthFirstSearch/NumberOfClosedIslands_1254.java
package algo.depthFirstSearch;

public class NumberOfClosedIslands_1254 {
    public static int closedIsland(int[][] grid) {
        var num = 0;
        var visited = new boolean[grid.length][grid[0].length];
        for (var i = 1; i < grid.length; i++) {
            for (var j = 1; j < grid[0].length; j++) {
                if (grid[i][j] == 0 && !visited[i][j] && findIslands(grid, i, j, visited)) {
                    num++;
                }
            }
        }
        return num;
    }

    private static boolean findIslands(final int[][] grid, final int row, final int col, final boolean[][] visited) {
        int rows = grid.length, cols = grid[0].length;
        if (visited[row][col] || grid[row][col] == 1) {
            return true;
        }
        visited[row][col] = true;

        // we need to visit other cells even if we reach border so that we can mark all cells visited
        var reachBorder = row == 0 || row == rows - 1 || col == 0 || col == cols - 1;

        var directions = new int[][]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        var result = true;
        for (var direction : directions) {
            int nextRow = row + direction[0], nextCol = col + direction[1];
            if (nextRow >= 0 && nextRow < rows && nextCol >= 0 && nextCol < cols) {
                if (!findIslands(grid, nextRow, nextCol, visited)) {
                    // we can not return from here as we need to visit all cells
                    result = false;
                }
            }
        }
        return !reachBorder && result;
    }

    public static void main(String[] args) {
        System.out.println(closedIsland(new int[][]{{1,1,1,1,1,1,1,0},
                                                    {1,0,0,0,0,1,1,0},
                                                    {1,0,1,0,1,1,1,0},
                                                    {1,0,0,0,0,1,0,1},
                                                    {1,1,1,1,1,1,1,0}}));
    }
}

File: src/main/kotlin/algo/depthFirstSearch/NumberOfClosedIslands_1254.kt
package algo.depthFirstSearch

class KtNumberOfClosedIslands_1254 {
    fun closedIsland(grid: Array<IntArray>): Int {
        fun bfs(grid: Array<IntArray>, row: Int, col: Int, visited: Array<BooleanArray>): Int {
            val cell = grid[row][col]
            if (visited[row][col] || cell == 1) {
                return 1
            }
            val rows = grid.size; val cols = grid[0].size
            visited[row][col] = true
            val reachBorder = row == rows - 1 || col == cols - 1 || row == 0 || col == 0
            var count = 0
            if (row + 1 < rows) {
                count += bfs(grid, row + 1, col, visited)
            }
            if (row > 0) {
                count += bfs(grid, row - 1, col, visited)
            }
            if (col + 1 < cols) {
                count += bfs(grid, row, col + 1, visited)
            }
            if (col > 0) {
                count += bfs(grid, row, col - 1, visited)
            }
            return if (reachBorder || count < 4) 0 else 1
        }

        var num = 0
        val rows = grid.size; val cols = grid[0].size
        val visited = Array(rows) { BooleanArray(cols) }
        grid.indices.forEach { row ->
            grid[0].indices.forEach { col ->
                if (grid[row][col] == 0 && !visited[row][col]) {
                    num += bfs(grid, row, col, visited)
                }
            }
        }
        return num
    }
}